Submenu nesting and saving new nested items.

tellthemachines opened the pull request:
## Description
<!-- Please describe what you have changed or added -->

Closes #21590 and #21593

Added ability to render submenus as nested, and save nested items, in the experimental navigation page.

TODO:  Render new nested items in the correct order (currently they're appended to the end of the list).

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Tested in browser.

## Screenshots <!-- if applicable -->

<img width="1085" alt="Screen Shot 2020-04-17 at 5 27 03 pm" src="https://user-images.githubusercontent.com/8096000/79543429-c97a2680-80d0-11ea-8962-95ef3bd26581.png">

We might need some styling fixes for this 🙀 :

<img width="1028" alt="Screen Shot 2020-04-17 at 5 27 26 pm" src="https://user-images.githubusercontent.com/8096000/79543466-d860d900-80d0-11ea-91ba-fdf2ddc1df58.png">


## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

tellthemachines committed:
Submenu nesting and saving new nested items.

talldan reviewed (commented):
This looks good, but It'd be great to see some helpful comments as the code is significantly more complicated now.

I also think considering how it can be unit tested soon would be good given the complexity.

noisysocks commented on packages/edit-navigation/src/components/menu-editor/use-navigation-blocks.js:
```diff
@@ -41,46 +46,76 @@ export default function useNavigationBlocks( menuId ) {
-			menuItemsRef.current[ block.clientId ] = menuItem;
-			innerBlocks.push( block );
-		}
+		const createMenuItemBlocks = ( items ) => {
```
We're going to want to unit test as much of this logic as possible. It's probably easiest to do that if we move the logic that turns a flat map of menu items into a tree of blocks out of the React hook and into a pure function that is name-exported.

```js
function createBlocksFromMenuItems( menuItems ) {
	...
}
...
const { blocks, menuItemsByClientId } = createBlocksFromMenuItems( menuItems );
menuItemsRef.current = menuItemsByClientId;
setBlocks( [ createBlock( 'core/navigation', {}, blocks ) ] );
```

This is kind of what I had in mind when I wrote `createBlockFromMenuItem`, but I didn't really think it through properly.

Happy to defer this for when we do https://github.com/WordPress/gutenberg/issues/21285.

noisysocks commented on packages/edit-navigation/src/components/menu-editor/use-navigation-blocks.js:
```diff
@@ -41,46 +46,76 @@ export default function useNavigationBlocks( menuId ) {
+
+		const saveNestedBlocks = ( nestedBlocks, parentId ) => {
+			for ( const block of nestedBlocks ) {
+				if ( block.innerBlocks && block.innerBlocks.length > 0 ) {
```
I think that `createBlock()` always sets `innerBlocks`, so checking for that isn't necessary.

```js
if ( block.innerBlocks.length ) {
```

noisysocks commented on packages/edit-navigation/src/components/menu-editor/use-navigation-blocks.js:
```diff
@@ -41,46 +46,76 @@ export default function useNavigationBlocks( menuId ) {
+				const menuItem = menuItemsRef.current[ block.clientId ];
+				const parentItemId =
+					menuItemsRef.current[ parentId ] &&
+					menuItemsRef.current[ parentId ].id;
```
We can use `?.` to help us out here.

```js
const parentItemId = menuItemsRef.current[ parentId ]?.id;
```

noisysocks commented on packages/edit-navigation/src/components/menu-editor/use-navigation-blocks.js:
```diff
@@ -41,46 +46,76 @@ export default function useNavigationBlocks( menuId ) {
+					saveMenuItem( {
+						...menuItem,
+						...createMenuItemAttributesFromBlock( block ),
+						menus: menuId, // Gotta do this because REST API doesn't like receiving an array here. Maybe a bug in the REST API?
```
Do we need to set `parent` here? What happens if a Navigation Link is moved from one submenu to another?

tellthemachines replied:
Good question; it doesn't appear to be possible to move items between nesting levels at the moment. Also, I just realised that nav items can be _dragged_ by the handle from the top toolbar and the experience is... not delightful 🙀

draganescu replied:
We should still send parent even if our UI is missing proper nesting, I think!

tellthemachines commented on packages/edit-navigation/src/components/menu-editor/use-navigation-blocks.js:
```diff
@@ -41,46 +46,76 @@ export default function useNavigationBlocks( menuId ) {
 		}
 
-		menuItemsRef.current = {};
+		const itemIndex = {};
+		menuItems.forEach( ( item ) => {
+			if ( ! itemIndex[ item.parent ] ) {
+				itemIndex[ item.parent ] = [];
+			}
+			itemIndex[ item.parent ].push( item );
+		} );
```
Ideally we would point to the [API reference](https://developer.wordpress.org/rest-api/reference/) for the structure of `menuItems`, but the menu endpoint isn't documented yet. I'll just say it's an array of menu item objects for now.

tellthemachines replied:
> Could maybe use lodash's groupBy? 

💡

tellthemachines commented on packages/edit-navigation/src/components/menu-editor/use-navigation-blocks.js:
```diff
@@ -41,46 +46,76 @@ export default function useNavigationBlocks( menuId ) {
-		}
+		const createMenuItemBlocks = ( items ) => {
+			const innerBlocks = [];
+			for ( const item of items ) {
```
I went for `items` here because eslint whinges about `menuItems` being already declared in the upper scope.

tellthemachines committed:
Address review feedback

tellthemachines commented:
I've addressed all the feedback so ready for another round of review! 

Regarding saving the menu order, it doesn't seem to be too straightforward so it's probably better to do it in a follow up PR.

tellthemachines committed:
Save parent items before nested items

tellthemachines committed:
Save sub-items of new items

tellthemachines committed:
Remove logging in error.

tellthemachines commented on lib/class-wp-rest-menu-items-controller.php:
```diff
@@ -446,6 +446,7 @@ protected function prepare_item_for_database( $request ) {
 						$prepared_nav_item['menu-item-position'] = count( $menu_items );
 					}
+					array_push($menu_items,$last_item);
 				} else {
```
I had to add this in order to be able to save sub-menu items nested under newly-created items. The problem is that the last item is being removed from `$menu_items` on line 443, and `$menu_items` is then used to create a `$menu_item_ids` array that is used to check if an id is a valid parent. The id of the last saved item will never appear in that array because it was popped out of `$menu_items`. 

This solution feels a bit hacky though; ideas for improvement appreciated 🙂

tellthemachines commented on lib/class-wp-rest-menu-items-controller.php:
```diff
@@ -446,6 +446,7 @@ protected function prepare_item_for_database( $request ) {
 						$prepared_nav_item['menu-item-position'] = count( $menu_items );
 					}
+					array_push($menu_items,$last_item);
 				} else {
```
I had to add this in order to be able to save sub-menu items nested under newly-created items. The problem is that the last item is being removed from $menu_items on line 443, and $menu_items is then used to create a $menu_item_ids array that is used to check if an id is a valid parent. The id of the last saved item will never appear in that array because it was popped out of $menu_items.

This solution feels a bit hacky though; ideas for improvement appreciated 🙂

draganescu replied:
The question is: does `$menu_items` at line 447 need its last item removed? If not, we can get the last item without removing it from the array:

```PHP
$last_item = $menu_items[count($menu_items)-1];
```

and use it on line 445, and if we need one less item on line 447 (for some reason) we'd be better off subtracting 1 there.

I wouldn't say your solution is hacky, just that maybe `array_pop` followed by an `array_push` has the risk of new code appearing between them which is unaware of these updates to `$menu_iteems`.

tellthemachines replied:
Thanks @draganescu , that's definitely better! We do need one less item on line 447, but agree it's better to subtract 1.

tellthemachines committed:
Array manipulation improvement

adamziel commented:
I just want to surface the discussion about saving menu items in a transactional way here:

https://github.com/WordPress/gutenberg/issues/21964#issuecomment-623310686

tellthemachines commented on packages/edit-navigation/src/components/menu-editor/use-navigation-blocks.js:
```diff
@@ -89,7 +124,7 @@ export default function useNavigationBlocks( menuId ) {
 		// Disable reason, this code will eventually be implemented.
 		// eslint-disable-next-line no-unused-vars
-		for ( const clientId of deletedClientIds ) {
+		for ( const deletedClientId of deletedClientIds ) {
```
If I do that, it fails the lint check because `clientId` is already declared in the upper scope. The deletion work will require some reworking to integrate with this PR anyway - which is what I attempted in #22128 😅

tellthemachines commented on packages/edit-navigation/src/components/menu-editor/use-navigation-blocks.js:
```diff
@@ -41,46 +47,75 @@ export default function useNavigationBlocks( menuId ) {
+						...createMenuItemAttributesFromBlock( block ),
+						menus: menuId,
+						parent: parentId,
+					} ).then( ( result ) => result );
```
I added that when I was trying to save reordering, where I needed to know the parent's position in the menu so I could order its children. That turned out to not be possible with the current API endpoints, and I forgot to remove this code.

tellthemachines committed:
Address review feedback

noisysocks commented:
There are some PHP lint errors which is what is causing Travis CI to fail. 

```
npm run lint-php
```

noisysocks reviewed (approved):
Works great! Nice work. Just fix up those PHP lint errors and we're good to go 👍 :shipit:

tellthemachines committed:
Fix PHP lint errors
